feat: show last viewed challenge in curriculum map on reload

ojeytonwilliams opened the pull request:
Previously, when the browser was reloaded, the curriculum map returned to the default state - it would open Responsive Web Design Certification > Basic HTML and HTML5.

This changes that behaviour, expanding the map to show the last challenge viewed.  For example, if they had been on /learn/javascript-algorithms-and-data-structures/es6/explore-differences-between-the-var-and-let-keywords last, then the curriculum will expand JavaScript Algorithms and Data Structures Certification > ES6 

Closes  #34236

ojeytonwilliams committed:
feat: Curriculum page loads with current challenge block open

Bouncey commented on client/src/components/Map/components/SuperBlock.js:
```diff
@@ -79,21 +79,48 @@ export class SuperBlock extends Component {
+    );
+    // If, somehow, the user does not have a current challenge, this opens the
+    // first superBlock.
+    const hasCurrentChallenge = noCurrentChallenge
```
I don't understand this. You are using a `boolean` condition in a ternary to resolve a `boolean`?

Could you maybe describe what you are trying to achieve here?

ojeytonwilliams replied:
Do you mean what is `isOpen` for?  What I was trying to achieve was to get a block to be expanded if the current challenge is in it, but it also needs to respond to user clicks.

Originally I was worried that you would get weird behaviour in an edge case, but as I was writing out my explanation, I convinced myself it wasn't a problem.  So, I was wrong and booleans are fine after all!

Bouncey commented on client/src/components/Map/components/SuperBlock.js:
```diff
@@ -79,21 +79,48 @@ export class SuperBlock extends Component {
+          challenge => challenge.id === currentChallengeId
+        );
+    const isOpen =
+      typeof isExpanded === 'undefined' ? hasCurrentChallenge : isExpanded;
```
You receive `undefined` here because you have adjusted the `expandedSelector`. It used to return a `boolean` at all times.

Bouncey commented on client/src/components/Map/components/SuperBlock.js:
```diff
@@ -79,21 +79,48 @@ export class SuperBlock extends Component {
+      <li className={`superblock ${isOpen ? 'open' : ''}`}>
+        <div
+          className='map-title'
+          onClick={() => toggleSuperBlock(superBlock, !isOpen)}
```
`toggleSuperBlock` didn't need to know the current expanded state before. I would be wary of this and ask why it needed to be changed.

Stateless is easier to maintain.

Bouncey commented on client/src/components/Map/components/SuperBlock.js:
```diff
@@ -79,21 +79,48 @@ export class SuperBlock extends Component {
         </div>
-        {isExpanded ? this.renderBlock(superBlock) : null}
+        {isOpen
+          ? this.renderBlock(blocksForSuperBlock, noCurrentChallenge)
```
Why did this method need to change?

Bouncey commented on client/src/components/Map/components/Block.js:
```diff
@@ -127,9 +134,16 @@ export class Block extends Component {
     });
+    // If, somehow, the user does not have a current challenge, this opens the
+    // first block.
+    const hasCurrentChallenge = noCurrentChallenge
```
See related comments for `superBlock.js`

Bouncey commented on client/src/components/Map/components/Block.js:
```diff
@@ -127,9 +134,16 @@ export class Block extends Component {
+    const hasCurrentChallenge = noCurrentChallenge
+      ? blockDashedName === 'basic-html-and-html5'
+      : challenges.some(challenge => challenge.id === currentChallengeId);
+    const isOpen =
```
See related comments for `superBlock.js`

Bouncey commented on client/src/components/Map/components/Block.js:
```diff
@@ -56,13 +59,15 @@ export class Block extends Component {
+        category: 'Map Block Click',
+        action: blockDashedName
+      });
+      return toggleBlock(blockDashedName, !isOpen);
```
`toggleBlock` didn't need to know the current expanded state before. I would be wary of this and ask why it needed to be changed.

Stateless is easier to maintain.

ojeytonwilliams commented on client/src/components/Map/components/Block.js:
```diff
@@ -20,8 +20,9 @@ const mapStateToProps = (state, ownProps) => {
     expandedSelector,
     userSelector,
-    (isExpanded, { completedChallenges = [] }) => ({
+    (isExpanded, { currentChallengeId, completedChallenges = [] }) => ({
```
Sure.  I hadn't noticed there were differences between the selectors, but this makes perfect sense.

ojeytonwilliams commented on client/src/components/Map/components/SuperBlock.js:
```diff
@@ -79,21 +79,48 @@ export class SuperBlock extends Component {
+    const blocksForSuperBlock = nodes.filter(
+      node => node.superBlock === superBlock
+    );
+    const noCurrentChallenge = !nodes.some(
```
Great, that makes life a lot easier!

ojeytonwilliams committed:
feat: Curriculum page expansion changes from review

ojeytonwilliams commented:
Right, that should cover it.  To answer the questions I didn't address specifically: after I finally realised that everything could be done with booleans while keeping things stateless, I ended up reverting a lot of my changes.  Hope this version's okay!

RandellDawson commented:
I just tested this PR locally and notice that if you open two sections and close the browser, when you come back it only leaves the first open (closest to the top).  Is that intended?  I was kind of thinking that both would be opened or better yet, the last one opened and touched vs. just the one closest to the top.

Bouncey commented:
@RandellDawson, the intended behaviour is for the map to be open at a users current challenge. I think tracking the map state across sessions would be confusing for the user.

Bouncey commented on client/src/components/Map/components/SuperBlock.js:
```diff
@@ -86,14 +86,33 @@ export class SuperBlock extends Component {
+    );
+    // If, somehow, the user does not have a current challenge, this opens the
+    // first superBlock.
+    const hasCurrentChallenge = !currentChallengeId
```
From what I can gather from this code and your comments, you are trying to determine is the users current challenge is inside this superBlock/block and expanding accordingly.

I can see a couple of pitfalls in this approach, whilst not wrong, they could be considered an anti-pattern.
- We use the `expandedState` to decide which block or superBlock should be expanded. Having individual components decide this is not desired.
- You are essentially using the same code across two components, just working on different data sets. This could and should be extracted out to have only one piece of code to maintain and test.

In the guide navigation, we use the `window.location.pathname` to toggle which parts of the guide navigation should be expanded. I think this is something we could use here also, but it will require some changes in the `get-session-user` api.

There is a utility function in the api that can take a challengeId and return a URI. You could use that api in the `/user/get-session-user` api call and return the resolved path in the `sessionMeta` property of the response.

This handler deals the the session user request:
https://github.com/freeCodeCamp/freeCodeCamp/blob/f687f804d830bdab56b1186ac7d27826ec309d43/api-server/server/boot/user.js#L37-L86

The challengeUrlResolver is here:
https://github.com/freeCodeCamp/freeCodeCamp/blob/f687f804d830bdab56b1186ac7d27826ec309d43/api-server/server/boot/challenge.js#L166-L191

`challengeurlResolver` usage example:

```js
const challengeUrlResolver = createChallengeUrlResolver(app);

const challengeUrl = challengeUrlResolver(`${user.currentChallengeId}`);

challengeUrl //? '/learn/superBlock/block/challenge.dashedName'
```

Example usage with `Observable`'s
```js
// the $ is an rxjs convention to indicate to the reader that this variable is an Observable
const challengeUrlResolver$ = Observable.fromPromise(
  challengeUrlResolver(user.currentChallengeId)
);

// the forkJoin method works like Promise.all, but for Observables
Observable.forJoin(
  foo$,
  bar$,
  baz$,
  (foo, bar, baz) => {
    // foo$, bar$, baz$ Observables have resolved and their output is available
    // in the same order they were passed in
    //
    // return something after doing stuff with foo, bar and baz
  }
)
```


If you store the resolved `challengeUrl` in the `sessionMeta` property of the response, you will then have it available on the client side to toggles superBlocks and blocks as needed, without searching through every challenge for a specific `challengeId`.

Bouncey reviewed (dismissed):
The solution I have proposed should help this feature to conform to the `redux` pattern and remove the 'heavy' work of searching through every `superBlock` and `block` for a `challengeId`

ojeytonwilliams committed:
feat: Curriculum page expansion changes from second review

ojeytonwilliams commented:
First off, I'll take a look at the merge conflict ASAP, but I wanted to get your thoughts before I ploughed on any further.

The basic idea is that, as you suggested, the `currentChallengeUrl` returned by the server initializes the expanded state of the curriculum.  It does this via the `Map` component.  The initialization being done there because it should only happen once while the page is loaded.  As the user can load the curriculum directly or come to it from another page, the user details may have already returned from the server or not.  As such, there needs to be a check in both `componentDidMount` and `componentDidUpdate`.  Since `componentDidMount` has no notion of previous state, it's necessary to check if any blocks have been expanded to avoid toggling them closed.  

The other option I considered was to simply open them, rather than checking if initialization has happened and toggling them if it hasn't.  That avoids having the check, but could be confusing if something a user was expecting to stay closed springs open of its own accord.

---
What remains?  As it stands, `getSessionUser` is only called once, so the `currentChallengeUrl` only updates once at the beginning.  If the user completes a block, starts a new one and then returns to the curriculum it will be expanded exactly as it was before they did the work.  To my mind this might feel like their work wasn't being registered.  I think it makes sense to have block completion fire off an action to close the finished block (and maybe super block) and open the next one, but I'm not 100% sure.

ojeytonwilliams commented:
@raisedadead I've taken another look at this and I think it's basically working, though it does need a final review.

To summarise:

1. If you open up (or reload) your browser and go to learn it will expand the curriculum to show you the challenges in the block you last looked at.
2. If you toggle some blocks in the curriculum, navigate away and navigate back the curriculum will still be the same as when you left it.

I was slightly concerned that people might be confused by 2. since, if they complete a block and go back to the curriculum, it won't have collapsed.  However, people might also find it odd that a block they'd left open went and closed by itself.

ValeraS reviewed (changes requested):
IMHO
1. expanded state should be local
1. init expanded state with a current challenge only in a `Map` constructor
1. use `currentChallengeId` to find which superblock and block to expand in a `Map` constructor.
1. update `currentChallengeId` in the store in the `currentChallengeSaga`

I don't like when expanded state changes on `componentDidUpdate` because of while user data is fetched I may want to expand some block, but when I'll decide to click status will change and I'll click on a wrong link.

@ojeytonwilliams could you rebase on the current master?

ojeytonwilliams commented:
@ValeraS I see that it could be annoying if the server is slow for some reason.  However, if we just rely on the store, I don't think we can solve the problem - since the store is recreated when the page is reloaded.

The only possibilities I can think of right now are
1. Prevent `initializeExpandedState` being called if the user has toggled something manually
2. Use a cookie to remember the current challenge

What do you think?

ValeraS commented:
> since the store is recreated when the page is reloaded

How about this: in this case, somehow propose the user to reveal the current challenge.

raisedadead commented:
> somehow propose the user to reveal the current challenge.

In all honesty, we want things to be simple. Its very difficult to explain users, if it's not obvious users will not get it in the first place. :(

ValeraS commented:
> > somehow propose the user to reveal the current challenge.
> 
> In all honesty, we want things to be simple. Its very difficult to explain users, if it's not obvious users will not get it in the first place. :(

Agree. Maybe show the loading status while we fetch user data?

How often this page is the first page for the user?
How often does the user reload this page during his session?

raisedadead commented:
How often this page is the first page for the user?

When the navigate from external places, news, forum etc.

How often does the user reload this page during his session?

If you mean a hard reload, that will be rarer. But the map is the only way to navigate the challenges. More so, when we remove it from the search soon.

raisedadead commented:
But, if we do have a nice way to do a loader and then load the component, I would be okay with storing the current challenge on the user object.

ojeytonwilliams commented:
As I see it there are two, more or less independent, issues.

1. What should happen when you navigate internally: should it remember the expanded state or recreate it based on the last challenge viewed?
2. What should happen when you reload: should it throw up a loader until the user state is loaded, show the curriculum immediately, but update it once the data arrives (assuming no user interaction) or use a cookie to track the last challenge viewed?

I don't have a strong feeling about 1, there are pros and cons to both.  However, I'm not keen on the loader idea, simply because we don't *need* the user data to show them the curriculum.  Even under normal conditions it takes a second or two before the user data comes back.  I don't think people like being made to wait.

I think you'd get a better user experience if you just used the local state if it exists, a cookie if it doesn't and default to having the first block open otherwise.

raisedadead commented:
> I think you'd get a better user experience if you just used the local state if it exists, a cookie if it doesn't and default to having the first block open otherwise.

Yes, I think this is going to be the best experience.

ojeytonwilliams committed:
fix: use a cookie to track current challenge

ojeytonwilliams commented:
If we go with that, presumably we can just ditch all the server side stuff.  I guess it's not the end of the world if someone deletes their cookies and ends up seeing the default expansion.

raisedadead commented:
> If we go with that, presumably we can just ditch all the server side stuff. I guess it's not the end of the world if someone deletes their cookies and ends up seeing the default expansion.

That is correct.

ValeraS commented:
I think, in most cases, a user goes through challenges one by one in order as it is and not often return to the Map. The Map is a start point, and if we will not change behavior on the first enter wy change it at all?

When you use FreeCodeCamp on two computers, a solution with the local state(cookie, localStorage, etc) for the first enter does not work well.

ojeytonwilliams committed:
fix: remove redundant server changes

ojeytonwilliams commented:
Sure, this won't fix all the possible problems, but it does help with the problems in #34236.

The only thing I can think of we could change to deal with, say, two computers, is to use the server's response to set a cookie.  However, that would only help if people logged in on a non-curriculum page and then navigated to learn.  If that's a common enough use case, I'd be happy to add that.

ojeytonwilliams committed:
fix: add PropTypes to fix lint error

ValeraS commented on client/src/components/Map/Map.js:
```diff
@@ -1,14 +1,27 @@
 import PropTypes from 'prop-types';
 import uniq from 'lodash/uniq';
+import { createSelector } from 'reselect';
+import cookies from 'browser-cookies';
```
Cookies are sent with every request to the server, but we do not need this. Use `LocalStorage` instead for this task.

ojeytonwilliams replied:
Good point. I thought you could just store cookies locally, but I was wrong.  I'll change it to use local storage instead.

ValeraS commented on client/src/components/Map/Map.js:
```diff
@@ -18,10 +31,57 @@ const propTypes = {
+      props.currentChallengeUrl ||
+      props.nodes[0].fields.slug;
+
+    if (!this.props.isInitialized)
```
I still don't see why should we persist expanded state during the session?

ojeytonwilliams replied:
It's a bit of an edge case, but that is a fallback if the user disabled cookies (or local storage since we're going to use that).  It won't matter often.

ValeraS replied:
I went through lessons, then return to the Map, and I'll get state like I have not done anything.

ojeytonwilliams replied:
Yeah, that's because I wasn't sure if it was better to expand the Map to the current challenge each time the user goes there or only when they reload the page.  It's easy enough to change if that's the desired behaviour.

ValeraS replied:
@raisedadead besides this, LGTM

raisedadead replied:
I would prefer if we expand each time user went there.

ValeraS replied:
What do you mean, expand the current challenge or last state?

ojeytonwilliams replied:
@ValeraS @raisedadead This latest commit should do the trick.

ValeraS replied:
It's the trick the right solution will be using local state.

raisedadead replied:
> I would prefer if we expand each time user went there.

I would prefer to expand to whatever last challenge I was on before clicking the `learn` (Map) on the nav, with a fallback to whatever we have in the state locally.

ojeytonwilliams replied:
I understood @raisedadead's comment
> I would prefer if we expand each time user went there.

to mean he wanted the `Map` to expand to show the last viewed challenge whenever you visited it, rather than just after reloading the page.  Which is now what happens.

> It's the trick right solution will be using local state.

It *does* use local state; it uses `currentChallengeId`.

ValeraS replied:
It uses the store for currently open sections but should use local state.

ojeytonwilliams replied:
The blocks and superblocks need to be able to be able to manipulate the state, which is rather easy if they interact with the store.  What's wrong with doing it that way?

Also, the point of this PR was to add a feature.  If we want to further refactor things, why not do that in a separate PR?

ValeraS replied:
I only said it's a trick because now you need to reset state in the store on each creation. By the logic, it's a local state not global. if you want, we can fix it in another PR.

ojeytonwilliams replied:
Ah, I see what you mean.  I think 'does the trick' was a poor choice of phrase by me.  It's an idiom meaning "achieves the goal".  Anyways, I see why resetting it is kind of a trick.  Perhaps we can come up with a nice, local version, later.

ValeraS commented on client/src/components/Map/Map.js:
```diff
@@ -1,14 +1,27 @@
+import { blockNameify } from '../../../utils/blockNameify';
 
 const propTypes = {
+  currentChallengeId: PropTypes.string,
```
I do not see the use of `currentChallengeId`, but I think that it should be used.

ojeytonwilliams replied:
That was used in the first version of this PR, so you could search for the challenge's URL.  It's definitely not necessary any more, though, since we'll be using local storage or just passing the `challengeUrl` directly. 

I'll just take it out.  Unless I'm missing something!

ValeraS replied:
We get `currentChallengeId` from a user data, it can be more actual than `challengeUrl`.

ojeytonwilliams replied:
If you mean more recent, I'm not sure that's true.  As far as I can see, user data doesn't get refreshed unless you reload the page.  Also, wasn't the idea to use local storage instead? After all, updating based on the server response can cause the page to change while the user tries to click on something.

raisedadead replied:
I think we are not on the same page here. I just went through the conversation again.

The `currentChallengeId` is the last challenge that the user has submitted successfully and is on the user records. It is meant to be updated only when user clicks the submit button (and passes the challenge correctly).

We need not update that key anytime else. That is by design.

The navigation state can be anything else and should be local to the session. We should not need to care about syncing sessions (rare edge case).

This does bring some ambiguity on how to gauge user intent on the each time the map loads in the session. I think I like the idea of simply checking the localStorage / cookie and expanding the block if it is present and if not opening up the last challenge they submitted based on currentChallengeId, and finally falling back to the default of the first block.

raisedadead replied:
Apologies about the confusion. In all honesty, I hope we should have named the key on user as `currentlySolvedChallengeId` !! But DB migration is a huge nightmare.

ValeraS replied:
You are not correct here, `currentChallengeId` is updated for registered users on a mount action of any challenge:

https://github.com/freeCodeCamp/freeCodeCamp/blob/fcfbe0314f4999ea3d57e5aa7309044c6d1eeb3e/client/src/templates/Challenges/redux/current-challenge-saga.js#L53

raisedadead replied:
No I agree, that is updated every time. 

But I don't think that is intended the design. What you have found is actually a bug. 

The intent I guess was to store the successfully submitted challenge as their current challenge. But then it got 're-used' to store the navigation as well 🤷‍♂ !

So now we already have `completedChallengeId` and `currentChallengeId`

ojeytonwilliams replied:
How about `lastViewedChallengeId`?  Something completely different, so we never mix them up.

ojeytonwilliams replied:
Actually, if we're using it for navigation, we'd use `lastViewedChallengeUrl`, but that general idea.

raisedadead replied:
Well, I would not want to rename or add a key to the User Object. DB migration is a nightmare. If we are able to achieve the intent with acceptable fallbacks, then that would be ideal.

ValeraS replied:
I don't see, why do we need `completedChallengeId`, we always can get it from `completedChallenges`? `currentChallengeId` is where we were the last time.

ojeytonwilliams replied:
Sure, I meant the store key though, i.e. what I'm calling `currentChallengeUrl`, rather than anything server side.

Just to clear up a couple of things.  

Firstly, should this PR only change what happens when the user reloads the page?  I.e. if they just navigate around, complete challenges or whatever, the expanded state stays as they left it. 

Secondly, should we go back to having the server add `currentChallengeUrl` to `sessionMeta`?  Or is it enough to use the store, so that if they reload a challenge learn will expand that block?

The latest commit only re-initialises the expanded state on page reload and does not rely on the server.  Instead it puts the url in the store and uses that.  So, if they reload a challenge, then go back to the curriculum, that challenge's block will be expanded (assuming no local storage/cookie).

The fallback goes cookie (though I'll change that to local storage) -> store value -> default first block.

ValeraS replied:
As I see it:
- on challenge mount, additionally, put `currentChallengeId` in the store and the local storage;
- on the Map, take `currentChallengeId` from the store and if it is not set then from local storage;
- find by `currentChallengeId` SuperBlock and Block to expand;
- and maybe scroll to a current challenge.

raisedadead replied:
I think we can avoid that one trip to the server. Let's add a commit to use localStorage instead of cookies, see how this behaves (on .dev) and then evaluate our decision.

ojeytonwilliams committed:
fix: change cookies to store and add tests

ojeytonwilliams committed:
Merge branch 'master' into feat/open-on-current-challenge

ojeytonwilliams commented:
I deliberately haven't added tests for the `Show`s, since they have graphQL queries and that means we have to mock Gatsby which complicates things a bit.

Other than that, I've switched from cookies to store and changed the fallback order as Valeriy suggested.

@raisedadead There's more we could do, but I think this is in a reasonable state.

raisedadead commented:
Hi @ojeytonwilliams just for the sake of posterity, can update the OP to reflect what this PR does.

I am going to be pulling this soon for a QA.

Thanks.

ValeraS commented:
@ojeytonwilliams I still don't understand why do you not want to use the `currentChallengeId` instead of `currentChallengeUrl`?

```js
const { superBlock, block } = nodes.find(node => node.id === curentChallengeId) || {};
```

IMHO, the behavior should be the same, whenever we go to the Map, i.e. do not save a state of the Map in the store.

ojeytonwilliams commented:
The idea was to avoid having to search through every single challenge each time the browser is reloaded.  We can do that, of course, but I don't see why that's preferable to putting a url in the store.  

In general I'd prefer not re-write working code when I don't see much to be gained by doing so.  In this case, as you say, the behaviour would be the same, so why change it?

ValeraS commented:
Because you do not use information from the user data what.

https://github.com/freeCodeCamp/freeCodeCamp/pull/35421#discussion_r312576332
> We get `currentChallengeId` from a user data, it can be more actual than `challengeUrl`.

If you do not want to use `currentChallengeId`, then putting` currentChallengeUrl` in the store is pointless in the current implementation.

raisedadead commented:
From https://github.com/freeCodeCamp/freeCodeCamp/pull/35421#discussion_r317678168

> - find by `currentChallengeId` SuperBlock and Block to expand;

Isn't this an expensive lookup? 

Hence I said, to break the deadlock on this PR:

> ... see how this behaves (on .dev) and then evaluate our decision.

ValeraS commented:
> Isn't this an expensive lookup?
> 
> Hence I said, to break the deadlock on this PR:

It will be done only on mount and `nodes` array contains about 1500 elements. it's not so many.

raisedadead commented:
> It will be done only on mount and nodes array contains about 1500 elements. it's not so many.

Agreed. And we can test this on our beta site? I have found the calls to API slow as it is :(

ojeytonwilliams committed:
fix: use currentChallengeId

ojeytonwilliams committed:
fix: update tests

I couldn't figure out how to test the challenge saga, so that's gone for
now.  The Map tests have been updated to use currentChallengeId

ojeytonwilliams committed:
fix: remove unused PropTypes

ojeytonwilliams commented:
There we go.  @ValeraS, @raisedadead it should do what we want, but using `currentChallengeId` and `componentDidMount` rather than the constructor.  

I changed it so that only signed in users can update any stores,  Since you can't make progress while logged out, it seemed inconsistent to allow the stores to change.  Unfortunately, this meant the saga became harder to test, so if anyone knows the right way to do that, I'd love to see it.

ValeraS commented:
Let's make one another field `currentChallengeId` in the store, independent from the user data with the next behavior:
- initialized with data from local storage or null;
- updated on `challengeMounted` and `fetchUserComplete` actions;
- used on the Map creation.

@ojeytonwilliams @raisedadead What do you think?

RandellDawson commented:
Out of curiosity, have you all considered tackling the issue reported in #https://github.com/freeCodeCamp/freeCodeCamp/issues/17778 while working on changing this map feature?

ojeytonwilliams committed:
fix: separate currentChallengeId from user

ojeytonwilliams commented:
@ValeraS Makes sense to me.  `currentChallengeId` is now its own property.  I went with defaulting to null, since the store is there as a fallback. I also fixed a bug in the saga.

@RandellDawson while I think that issue needs dealing with (amongst other accessibility problems), I think we should try to keep PRs focused, if possible.

ValeraS commented on client/src/components/Map/Map.js:
```diff
@@ -18,10 +27,50 @@ const propTypes = {
+export class Map extends Component {
+  componentDidMount() {
+    const currentChallengeId =
+      this.props.currentChallengeId || store.get(CURRENT_CHALLENGE_KEY);
```
```suggestion
      this.props.currentChallengeId;
```

ValeraS commented on client/src/redux/index.js:
```diff
@@ -28,6 +28,7 @@ export const defaultFetchState = {
 const initialState = {
   appUsername: '',
   completionCount: 0,
+  currentChallengeId: null,
```
Still, think it should be initialized from local storage.

ValeraS commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -16,14 +18,20 @@ import { post } from '../../../utils/ajax';
   const isSignedIn = yield select(isSignedInSelector);
   const currentChallengeId = yield select(currentChallengeIdSelector);
-  if (isSignedIn && payload !== currentChallengeId) {
+  // TODO: should the server update take place on challenge completion instead?
```
Do we need it?

ojeytonwilliams replied:
Not sure, that's why I left the TODO.  @raisedadead?

raisedadead replied:
Yes, this is the note that I was referring to our design of the system earlier. We intended to update the challenge update only on successful completion. The challenge navigation should have been kept different from this. 

Anyway, since we are taking a new direction, lets remove this note.

ValeraS reviewed (changes requested):
I don't see why a behavior for a logged user should be different from the behavior for a not logged user in this case?

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -16,14 +18,20 @@ import { post } from '../../../utils/ajax';
+  // TODO: should the server update take place on challenge completion instead?
+  if (isSignedIn && id !== currentChallengeId) {
+    store.set(CURRENT_CHALLENGE_KEY, id);
+    yield put(updateCurrentChallengeId(id));
```
`challengeMounted` doesn't have access to the same `state` as `fetchUserComplete`, does it?  Is there a good way to communicate between the independent `state`s other than by firing off a second action?

ValeraS replied:
All reducers receive all actions. Add `challengeMounted` action processing in the main reducer.

ojeytonwilliams replied:
Okay, so doing it this way means that 
```js
id !== currentChallengeId
```
is always false, because `challengeMounted` will have already updated `currentChallengeId` by this point.  Hence the server never gets updated.

I see two options.
1) Remove `id !== currentChallengeId` and accept that the server is going to get more unnecessary requests.
2) Use `updateCurrentChallengeId` as before.

Unless there's a third option that I'm missing, which do you think is best?

ValeraS replied:
Currently, we can safely remove this check, it did not work properly, because the `currentChallengeId` was not updated during the session.

ojeytonwilliams replied:
I was worried it would ping the server each time you reloaded a challenge.  However, that should not happen because the challenge gets mounted before the user comes back from the server.  i.e. `isSignedIn` is false, so the second check doesn't matter.

ojeytonwilliams commented on client/src/templates/Challenges/redux/current-challenge-saga.js:
```diff
@@ -16,14 +18,20 @@ import { post } from '../../../utils/ajax';
-  if (isSignedIn && payload !== currentChallengeId) {
+  // TODO: should the server update take place on challenge completion instead?
+  if (isSignedIn && id !== currentChallengeId) {
+    store.set(CURRENT_CHALLENGE_KEY, id);
```
Just to clarify, you think 
```js
yield call(post, update.endpoint, update.payload);
```
should happen even if you're not signed in?  Or did you just mean that
```js
store.set(CURRENT_CHALLENGE_KEY, id);
```
should happen for all users?

ValeraS replied:
The last one

ojeytonwilliams committed:
fix: update currentChallengeId directly on mount

ojeytonwilliams commented:
@ValeraS I've made the changes and re-checked the behaviour. Let me know if you spot anything else.

ojeytonwilliams committed:
feat: reset map on every visit

raisedadead commented:
@ValeraS please merge if you are happy with the changes. I would deploy this for a QA on the dev instance.

ojeytonwilliams commented:
Thanks for all the reviews @ValeraS.  I'm pretty happy with how it turned out.  Pending .dev QA, of course!

raisedadead commented:
Thanks @ojeytonwilliams @ValeraS @RandellDawson ! 🎉 

This was quite a tough feature to implement to be honest. I have triggered a build and it should be available within the hour.

> #### Thanks everyone for all your hard work!
